@here/harp-geometry/lib/TechniqueProcessor.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getPropertyValue } from "@here/harp-datasource-protocol";
+import { MapEnv } from "@here/harp-datasource-protocol/index-decoder";
+import * as THREE from "three";
+
+/**
+ * State of the scene where objects generated by a [[TechniqueProcessor]] live. Can be used for
+ * dynamic update of the object properties (uniforms, blending state, etc.).
+ */
+export interface SceneState {
+    readonly camera: THREE.PerspectiveCamera;
+    readonly zoomLevel: number;
+    readonly pixelToWorld: number;
+}
+
+/**
+ * Base interface for all [[Technique]]-related configuration paramets.
+ */
+export interface TechniqueParams {
+    /**
+     * [[Technique]] identifier.
+     */
+    readonly name: string;
+    /**
+     * The render order of the objects created using this technique.
+     *
+     * If not specified in style file, [[StyleSetEvaluator]] will assign monotonically increasing
+     * values according to style position in file.
+     */
+    readonly renderOrder: number;
+}
+
+/**
+ * Base class that handles processing data accordingly to its assigned [[Technique]].
+ */
+export abstract class TechniqueProcessor {
+    /**
+     * [[Technique]] identifier.
+     */
+    static readonly id: string = "none";
+
+    /**
+     * Array of uniforms assigned to the objects created with this [[TechniqueProcessor]].
+     */
+    static readonly uniforms: Map<string, THREE.Uniform>;
+    /**
+     * Vertex shader source code.
+     */
+    static readonly vertexShader: string;
+    /**
+     * Fragment shader source code.
+     */
+    static readonly fragmentShader: string;
+
+    /**
+     * Checks if this [[TechniqueProcessor]] can process the specified data.
+     *
+     * @param data Data to check for support.
+     */
+    static supportsData(data: ArrayBufferLike | {}): boolean {
```

Comment:
I agree, but currently cannot think of anything better (and to be honest, is the same types we use in the ``WorkerDecoderProtocol``...). Any suggestion? It's tricky to be flexible and strict at the same time.

Reply:
If this is a bare "drawing" technique intended at _just_drawing_lines_ i would think that they should accept world coordinates, rather bar buffers.

I think that mere reference to `WorkerDecoderProtocol` shows misunderstanding (possibly mine, but still communication issue) of purpose this task.

Advertised as "refactor-out" bare technique processing, now you're reusing interface of "tile decoder" (`ArrayBufferLike | {}` is basically MVT or MvtJson or GeoJSON) so, it's not layer we would like to interface with technique.

So, on which level technique API should be defined ? Some aspects.
 - should technique know/understand transfer formats? no, it should accept bare polygons/lines/points in some space
 - should technique know/understand projections? no, it should work in eucludian space, like world/tile space, so space doesn't matter as long as it's affine-transformable to world space
 
Ergo -> 
We've already interfaces in `IGeometryProcessor.ts` which IMHO are best suitable for technique input - with one difference, they should use convey world-space compatible coordinates instead of geocoords.

Varia:
_if_ we would lay a line on curved area (terrain or sphere), it's caller responsibility (decoder in this case) to divide geometry and add proper z-offsets